README.md
## Levenburg-Marquardt fitting of SuperDARN auto-correlation functions (ACFs). **With NO ad hoc assumptions/conditions!**
[As described by Reimer et. al. (2018)](https://doi.org/10.1002/2017RS006450).

Fitted SuperDARN data with accurate error bars. Filter data based on fitted error:


![alt text](example/animated.gif)

Pictured above is 8 hours of fitted velocity data for the Saskatoon SuperDARN radar for none, 800 m/s, 400 m/s, 200 m/s, 100 m/s, and 50 m/s velocity error thresholds.


# Documentation
More detailed documentation, including usage examples, is availabe on readthedocs.

# Summary
This repository contains code to that can fit the SuperDARN rawacf files using an error weighted non-linear least-squares fitting algorithm called the First-Principles Fitting Methodology (FPFM). The details of this algorithm as described in [Reimer et. al. (2018)](https://doi.org/10.1002/2017RS006450). The real and imaginary components are fitted against a decaying complex sinusoidal model of the ACF. The Levenburg-Marquardt technique is used to minimize the chi-squared sum. No ad hoc conditions are utilized. The errors in the real and imaginary components of the ACF lag estimates are based on the work by [Reimer et. al. (2016)](https://doi.org/10.1002/2016RS005975). Self-clutter due to the multiple-pulse technique is accounted for using the Mean Power-based Self-clutter Estimator (MPSE) detailed in [Reimer and Hussey (2015)](https://doi.org/10.1002/2015RS005706).

# C Code
The C code builds against [RST](https://github.com/superdarn/rst), specifically depending on the C dmap library and the hdw.dat files. For both this software and RSTLite you will need to have zlib-dev installed. It also utilizes the [cmpfit library](https://www.physics.wisc.edu/~craigm/idl/cmpfit.html) written by Craig B. Marquardt. The codebase has been structured in the same way as the [fitacf3.0 code](https://github.com/SuperDARNCanada/fitacf.3.0) written by Keith Kotyk.

Once the RSTLite and zlib dependencies are installed, to compile the `LMFIT2` code, simply change into the C code directory and use the make command. The compiled binary will be in the bin directory.

## Sourcing RST .profile.bash
Remember to source the .profile.bash file in the RST directory. If you do not, `LMFIT2` will not work

## RST dependency
The C version of `LMFIT2` specifically relies on RST these components: `fitblk.h`,`fitdata.h`,`rtypes.h`,`dmap.h`,`rprm.h`,`radar.h`, and `rawdata.h`.

# Python Code
The python version of `LMFIT2` requires the numpy and lmfit python packages to be installed. The code also relies on the python version of the dmap library [(pydmap written by Keith Kotyk)](https://github.com/SuperDARNCanada/pydmap), which is included in this repository.

# Limitations and TO DO
Currently, this code only fits the ACF not the XCF, so no elevation angles are fit for yet. Also, only the exponential model of the ACF is fit. A version of the code that fits both gaussian and exponential and ACF and XCF is under development.

# Speed
The C code takes approximately 9.5 minutes for 1 hour worth of raw data. Python hasn't yet been speed tested. The C code could be sped up by using an analytic Jacobian matrix instead of relying on MPFIT to numerically estimate it.



HISTORY.rst
.. :changelog:

1.0 (2018-05-12)
++++++++++++++++++

- Initial full release of code since `Radio Science publication <https://doi.org/10.1002/2017RS006450>`_
- Included unit testing for C and python versions
- Some bug fixes to the python version to bring it up to date with C version


C/tests/test.bash
#!/bin/bash

# get path to make_lmfit2 binary
CWD=`pwd`
LMFIT2=${CWD:0:-6}/bin/make_lmfit2

# dmapdump command (assuming it's in $PATH!)
DMAPDUMP="dmapdump"

# path to expected.lmfit2
EXPECTED=${CWD}/expected.lmfit2
EXPECTEDDUMP=${CWD}/expected.dmapdump
# path to input rawacf file: test.rawacf
INPUT=${CWD}/test.rawacf
# path to test fitted file: output.lmfit2
OUTPUT=${CWD}/output.lmfit2
OUTPUTDUMP=${CWD}/output.dmapdump

# Now run the code and fit the test.rawacf file
echo -e "\nFitting test.rawacf file. This will take a few seconds..."
$LMFIT2 -new $INPUT > $OUTPUT
echo "... done!"
sleep 1
# Now compare the fitted file to expected file
echo "Now comparing output to expected using dmapdump..."
sleep 1
$DMAPDUMP $EXPECTED > $EXPECTEDDUMP
$DMAPDUMP $OUTPUT > $OUTPUTDUMP

#trim out the origin.time and origin.command lines since
#these are expected to be different
sed -i '/origin.time/d' $EXPECTEDDUMP
sed -i '/origin.command/d' $EXPECTEDDUMP
sed -i '/origin.time/d' $OUTPUTDUMP
sed -i '/origin.command/d' $OUTPUTDUMP

# finally use diff to see if files are different
echo -e "Output of a diff:\n"
diff $EXPECTEDDUMP $OUTPUTDUMP

echo -e "\nIf lines are empty, files are the same!"

sleep 1
echo -e "Cleaning up output files...\n"
rm $EXPECTEDDUMP $OUTPUTDUMP $OUTPUT

exit 0



python/tests/test.py
import os
import sys
import unittest

sys.path.append('.')

from lmfit2 import read_file, lmfit2_mp, write_file, main

from numpy.testing import assert_array_equal

class TestLmfit2(unittest.TestCase):

    def test_fitting_multiprocessing(self):
        print("Running multiprocessing test on 2 threads.")

        import multiprocessing as mp

        pool = mp.Pool(2)

        input_filename = 'tests/test.rawacf'
        output_filename = 'tests/output_mp.lmfit2'
        expected_filename = 'tests/expected.lmfit2'

        fitted_records = list()

        #first, read the file we want to fit
        records = read_file(input_filename)

        temp = [pool.apply_async(lmfit2_mp,args=(x,records[x])) for x in range(len(records))]

        fitted_records = [p.get() for p in temp]
        fitted_records.sort()
        fitted_records = [r[1] for r in fitted_records]

        write_file(output_filename, fitted_records)

        output_file = read_file(output_filename)
        expected_file = read_file(expected_filename)

        for item1,item2 in zip(output_file,expected_file):
            item1_key = item1.keys()
            item1_key.sort()
            item2_key = item2.keys()
            item2_key.sort()

            self.assertEqual(item1_key,item2_key)

            for key in item1_key:
                assert_array_equal(item1[key],item2[key])


    def test_fitting_singleprocessing(self):
        print("Running single processing test.")

        input_filename = 'tests/test.rawacf'
        output_filename = 'tests/output_sp.lmfit2'
        expected_filename = 'tests/expected.lmfit2'

        main(input_filename,output_filename)


        output_file = read_file(output_filename)
        expected_file = read_file(expected_filename)

        for item1,item2 in zip(output_file,expected_file):
            item1_key = item1.keys()
            item1_key.sort()
            item2_key = item2.keys()
            item2_key.sort()

            self.assertEqual(item1_key,item2_key)

            for key in item1_key:
                assert_array_equal(item1[key],item2[key])


if __name__ == '__main__':
    unittest.main()

    os.remove('tests/output_mp.lmfit2')
    os.remove('tests/output_sp.lmfit2')


C/src/fitting.c
/*
LMFIT2 least square fitting wrapper functions

Adapted by: Ashton Reimer
From code by: Keith Kotyk

ISAS
August 2016

*/

#include <math.h>
#include "fitting.h"
#include "preprocessing.h"
#include "leastsquares.h"
#include <stdio.h>

/* TODO, IMPLEMENT QUADRATIC ENVELOPE AND XCF FITTING */


/**
Helper function that calls the lmfit code with appropriate arguments
depending on whether we are fitting the ACF or the XCF.
*/
void do_LMFIT(llist_node range,PHASETYPE *phasetype,FITPRMS *fitted_prms){
	RANGENODE* range_node;
	range_node = (RANGENODE*) range;
    double lambda, mpinc;

    lambda = 299792458.0/(fitted_prms->tfreq*1000.0);
    mpinc = fitted_prms->mpinc;

	switch(*phasetype){
		case ACF:
		    /* Do exponential envelope fit */
			lmfit_acf(range_node->l_acf_fit,range_node->acf,lambda,mpinc,1,0);

            /* Do gaussian envelope fit */
/*			lmfit_acf(range_node->q_acf_fit,range_node->acf,lambda,mpinc,1,1); */
			break;

		case XCF:
		    /* Do both fits here, take phi0 with smallest error */
		    /* Do exponential envelope fit */
/*			lmfit_xcf(range_node->l_xcf_fit,range_node->xcf,lambda,mpinc,1,0); */

            /* Do gaussian envelope fit */
/*			lmfit_xcf(range_node->q_xcf_fit,range_node->xcf,lambda,mpinc,1,1); */
			break;
	}
}


/**
For a each range, fit the ACF
*/
void ACF_Fit(llist ranges,FITPRMS *fitted_prms){
	PHASETYPE acf = ACF;
	llist_for_each_arg(ranges,(node_func_arg)do_LMFIT,&acf,fitted_prms);
}


/**
For a each range, fit the XCF
*/
void XCF_Fit(llist ranges,FITPRMS *fitted_prms){
	PHASETYPE xcf = XCF;
	llist_for_each_arg(ranges,(node_func_arg)do_LMFIT,&xcf,&fitted_prms);
}

C/include/preprocessing.h
/*
ACF Processing main functions

Adapted by: Ashton Reimer
From code by: Keith Kotyk

ISAS
August 2016

*/

#ifndef _ACFPROC_H
#define _ACFPROC_H

#include "llist.h"
#include "leastsquares.h"
#include "fit_structures.h"

#define MIN_LAGS 3
#define ACF_SNR_CUTOFF 10.0
#define ALPHA_CUTOFF 2
#define FLUCTUATION_CUTOFF_COEFF 2

RANGENODE* new_range_node(int range, FITPRMS *fit_prms);
PWRNODE* new_pwr_node(int range, LAGNODE* lag, FITPRMS *fit_prms);
ACFNODE* new_acf_node(int range, LAGNODE* lag, FITPRMS *fit_prms);
SCNODE* new_sc_node(int range, LAGNODE* lag, FITPRMS *fit_prms);
PHASENODE* new_elev_node(int range, LAGNODE* lag, FITPRMS *fit_prms);
void free_range_node(llist_node range);

int compare_doubles(llist_node first,llist_node second);
int compare_ints(llist_node first,llist_node second);
int compare_lags(llist_node first,llist_node second);
bool pwr_node_eq(llist_node node,llist_node cmp);
bool acf_node_eq(llist_node node,llist_node cmp);
bool phase_node_eq(llist_node node,llist_node cmp);
bool sample_node_eq(llist_node node,llist_node cmp);
bool range_node_eq(llist_node node,llist_node cmp);
void print_acf_node(llist_node node,FILE* fp);

void print_node(llist_node node);
void print_range_node(llist_node node);
void print_lag_node(llist_node node);
void print_pwr_node(llist_node node,FILE* fp);
void print_scpwr_node(llist_node node,FILE* fp);

void mark_bad_samples(FITPRMS *fit_prms, llist bad_samples);
void filter_tx_overlapped_lags(llist_node range, llist lags, llist bad_samples);
double ACF_cutoff_pwr(FITPRMS *fit_prms);

void Determine_Lags(llist lags,FITPRMS *fit_prms);
void Fill_Range_List(FITPRMS *fit_prms, llist good_ranges);
void Fill_Data_Lists_For_Range(llist_node range,llist lags,FITPRMS *fit_prms);
void Filter_TX_Overlap(llist ranges, llist lags, FITPRMS *fit_prms);

bool Is_Converged(llist *ranges,double tol);
void Estimate_Self_Clutter(llist *ranges,FITPRMS *fit_prms);
void Estimate_First_Order_Error(llist *ranges,FITPRMS *fit_prms, double noise_pwr);
void Estimate_Re_Im_Error(llist *ranges,FITPRMS *fit_prms, double noise_pwr);


#endif


python/lmfit2.py

fitacf_types = {
 'radar.revision.major' : 'c',
 'radar.revision.minor' : 'c',
 'origin.code' : 'c',
 'origin.time' : 's',
 'origin.command' : 's',
 'cp' : 'h',
 'stid' : 'h',
 'time.yr' : 'h',
 'time.mo' : 'h',
 'time.dy' : 'h',
 'time.hr' : 'h',
 'time.mt' : 'h',
 'time.sc' : 'h',
 'time.us' : 'i',
 'txpow' : 'h',
 'nave' : 'h',
 'atten' : 'h',
 'lagfr' : 'h',
 'smsep' : 'h',
 'ercod' : 'h',
 'stat.agc' : 'h',
 'stat.lopwr' : 'h',
 'noise.search' : 'f',
 'noise.mean' : 'f',
 'channel' : 'h',
 'bmnum' : 'h',
 'bmazm' : 'f',
 'scan' : 'h',
 'offset' : 'h',
 'rxrise' : 'h',
 'intt.sc' : 'h',
 'intt.us' : 'i',
 'txpl' : 'h',
 'mpinc' : 'h',
 'mppul' : 'h',
 'mplgs' : 'h',
 'nrang' : 'h',
 'frang' : 'h',
 'rsep' : 'h',
 'xcf' : 'h',
 'tfreq' : 'h',
 'tfreq2' : 'h',
 'mxpwr' : 'i',
 'lvmax' : 'i',
 'fitacf.revision.major' : 'i',
 'fitacf.revision.minor' : 'i',
 'combf' : 's',
 'noise.sky' : 'f',
 'noise.lag0' : 'f',
 'noise.vel' : 'f',
 'ptab' : 'h',
 'ltab' : 'h',
 'pwr0' : 'f',
 'slist' : 'h',
 'nlag' : 'h',
 'qflg' : 'c',
 'gflg' : 'c',
 'p_l' : 'f',
 'p_l_e' : 'f',
 'p_s' : 'f',
 'p_s_e' : 'f',
 'v' : 'f',
 'v_e' : 'f',
 'w_l' : 'f',
 'w_l_e' : 'f',
 'w_s' : 'f',
 'w_s_e' : 'f',
}

fit_record_keys = [
	 'radar.revision.major',
	 'radar.revision.minor',
	 'origin.code',
	 'origin.time',
	 'origin.command',
	 'cp',
	 'stid',
	 'time.yr',
	 'time.mo',
	 'time.dy',
	 'time.hr',
	 'time.mt',
	 'time.sc',
	 'time.us',
	 'txpow',
	 'nave',
	 'atten',
	 'lagfr',
	 'smsep',
	 'ercod',
	 'stat.agc',
	 'stat.lopwr',
	 'noise.search',
	 'noise.mean',
	 'channel',
	 'bmnum',
	 'bmazm',
	 'scan',
	 'offset',
	 'rxrise',
	 'intt.sc',
	 'intt.us',
	 'txpl',
	 'mpinc',
	 'mppul',
	 'mplgs',
	 'nrang',
	 'frang',
	 'rsep',
	 'xcf',
	 'tfreq',
	 'mxpwr',
	 'lvmax',
	 'fitacf.revision.major',
	 'fitacf.revision.minor',
	 'combf',
	 'noise.sky',
	 'noise.lag0',
	 'noise.vel',
	 'ptab',
	 'ltab',
	 'pwr0',
	 'slist',
	 'nlag',
	 'qflg',
	 'gflg',
	 'p_l',
	 'p_l_e',
	 'p_s',
	 'p_s_e',
	 'v',
	 'v_e',
	 'w_l',
	 'w_l_e',
	 'w_s',
	 'w_s_e',
	 'sd_l',
	 'sd_s',
	 'sd_phi',
	 'x_qflg',
	 'x_gflg',
	 'x_p_l',
	 'x_p_l_e',
	 'x_p_s',
	 'x_p_s_e',
	 'x_v',
	 'x_v_e',
	 'x_w_l',
	 'x_w_l_e',
	 'x_w_s',
	 'x_w_s_e',
	 'phi0',
	 'phi0_e',
	 'elv',
	 'elv_low',
	 'elv_high',
	 'x_sd_l',
	 'x_sd_s',
	 'x_sd_phi']


#read rawacf file using pydmap and return a list of 
#dictionaries each containing a beam record
def read_file(filename):
    from dmap import parse_dmap_format_from_file

    return parse_dmap_format_from_file(filename)


#write the fitted file using pydmap from a list of 
#dictionaries each containing a beam record
def write_file(filename, records):
    from dmap import RawDmapWrite
    RawDmapWrite(records,filename,ud_types=fitacf_types)


#Estimate teh noise power from the 10 lowest power
#ranges.
def estimate_noise(power):
    import numpy as np
    p = list(power)
    p.sort()
    noise = np.mean(p[0:20])
    return noise


#First order error bars for magnitude and acf fitting
def first_order_errors(power,noise,clutter,nave):
    import numpy as np
    return (power+noise+clutter)/np.sqrt(nave)


#Errors in real and imaginary components of acfs
def acf_error(power,noise,clutter,nave,rho,rho_c):
    import numpy as np
    return (power+noise+clutter)*np.sqrt((1-rho**2)/2. + rho_c**2)/np.sqrt(nave)


#Magnitude component errorbars
def mean_mag_stats_from_gaus(re,im,re_error,im_error,rho_r,rho_i,K,num_samps=10000):
    import numpy as np
    factor = 2*rho_r*rho_i
    m = np.array([re,im])
    c = np.array([[re_error**2,factor*re_error*im_error],
                  [factor*re_error*im_error,im_error**2]])
    temp = np.random.multivariate_normal(m,c,size=num_samps)
    x = temp[:,0]
    y = temp[:,1]
    r = np.sqrt(x**2 + y**2)

    if r is not None:
        mean = np.mean(r)
        var = np.var(r)
    else:
        mean = np.nan
        var = np.nan
    return mean, var

#Figure out which lags are bad due to transmitter blanking
#the receivers.
def determine_tx_blanked(nrg,ltab,tp,tau,tfr,gate):
    import numpy as np

    #Calculate the lags and the pulse table
    lags = []
    ptab = []
    for pair in ltab:
        lags.append(pair[1]-pair[0])
        ptab.extend(pair)
    lags=list(set(lags))
    ptab=list(set(ptab))
    lags.sort()  
    ptab.sort()

    txs_in_lag={}
    for lag in lags:
        txs_in_lag[lag]=[]

    #number of ranges per tau
    tp_in_tau = tau/tp

    #determine which range gates correspond to blanked samples
    tx_times=[p*tp_in_tau for p in ptab]
    blanked_samples=[]
    for tx in tx_times:
        blanked_samples.extend([tx,tx+1])

    for lag in lags:
        for pair in ltab:
            if (pair[1] - pair[0] == lag):
                #which samples are we using?
                S1=tp_in_tau*pair[0]+gate + 2.*tfr/(tp)
                S2=tp_in_tau*pair[1]+gate + 2.*tfr/(tp)
                br=[]
                #check to see if the samples are blanked or not
                if S1 in blanked_samples:
                    br.append(S1)
                if S2 in blanked_samples:
                    br.append(S2)
                txs_in_lag[lag].extend(br)

    return txs_in_lag

#Estimate upper limit of self clutter using the MPSE of 
#Reimer and Hussey 2015 (radio science)
def estimate_selfclutter(nrg,ltab,tp,tau,tfr,gate,power):
    import numpy as np

    #Calculate the lags and the pulse table
    lags = []
    ptab = []
    for pair in ltab:
        lags.append(pair[1]-pair[0])
        ptab.extend(pair)
    lags=list(set(lags))
    ptab=list(set(ptab))
    lags.sort()  
    ptab.sort()

    ranges_in_lag={}
    samples_in_lag={}
    for lag in lags:
        ranges_in_lag[lag]=[]
        samples_in_lag[lag]=[]

    #number of ranges per tau
    tp_in_tau = tau/tp

    for lag in lags:
        for pair in ltab:
            if (pair[1] - pair[0] == lag):
                #which samples are we using?
                S1=tp_in_tau*pair[0]+gate + 2.*tfr/(tp)
                S2=tp_in_tau*pair[1]+gate + 2.*tfr/(tp)
                samples_in_lag[lag].append([S1, S2])
                #which pulses were transmitted before the samples were recorded
                P1=[pulse for pulse in ptab if pulse*tp_in_tau <= S1]
                P2=[pulse for pulse in ptab if pulse*tp_in_tau <= S2]
                P1.sort()
                P2.sort()
                #what ranges are the previously transmitted pulses coming from
                r1 = [int(S1 - p*tp_in_tau - 2.*tfr/tp) for p in P1]
                r2 = [int(S2 - p*tp_in_tau - 2.*tfr/tp) for p in P2]
                r1.sort()
                r2.sort()
                ranges_in_lag[lag].append([r1, r2])

    cri_power = np.zeros((lags[-1]+1,))

    for l,lag in enumerate(lags):
        cris=[]

        S1=samples_in_lag[lag][0][0]
        S2=samples_in_lag[lag][0][1]
   
        #reimer method where the power profile is used to estimate the self-clutter
        #so that C is the geometric mean of the power from interfering ranges
        for range_pairs in ranges_in_lag[lag]:
            interfering_ranges = []
            #only use ranges that we have pwr0 for
            interfering_S1 = np.array([int(x) for x in range_pairs[0] if ((x != gate) & (x >= 0.) & (x < nrg-1))])
            interfering_S2 = np.array([int(x) for x in range_pairs[1] if ((x != gate) & (x >= 0.) & (x < nrg-1))])

            #First term in the summation for the self-clutter estimate
            term1 = 0
            if interfering_S1.size > 0:
                for ind in interfering_S1:
                    term1 += np.sqrt(power[gate]*power[ind])

            #Second term in the summation for the self-clutter estimate
            term2 = 0
            if interfering_S2.size > 0:
                for ind in interfering_S2:
                    term2 += np.sqrt(power[gate]*power[ind])

            #Third term in the summation for the self-clutter estimate
            term3 = 0
            if (interfering_S1.size > 0 and interfering_S2.size > 0):
                for i in interfering_S1:
                    for ind in interfering_S2:
                        term3 += np.sqrt(power[i]*power[ind])
            cris.append(term1 + term2 + term3)
        cri_power[lag] += cris[:]

    return cri_power


#Build the initial fit_record to write to file
def build_fit_record(record):
    import numpy as np

    fit_record = dict()
    key_list = ['radar.revision.major','radar.revision.minor','origin.code',
                'origin.time','origin.command','cp','stid','time.yr',
                'time.mo','time.dy','time.hr','time.mt','time.sc','time.us',
                'txpow','nave','atten','lagfr','smsep','ercod','stat.agc',
                'stat.lopwr','noise.search','noise.mean','channel','bmnum',
                'bmazm','scan','offset','rxrise','intt.sc','intt.us','txpl',
                'mpinc','mppul','mplgs','nrang','frang','rsep','xcf','tfreq',
                'mxpwr','lvmax','fitacf.revision.major','fitacf.revision.minor',
                'combf','noise.sky','noise.lag0','noise.vel','ptab','ltab','pwr0']
    for key in key_list:
        if key in record.keys():
            fit_record[key] = record[key]

    return fit_record


def acf_residual(params, time, re, im, re_error, im_error, lamda):
    import numpy as np
    pr = params['power']
    wd = params['width']
    vd = params['velocity']

    model1 = pr*np.exp(-time*wd*2*np.pi/lamda)*np.cos(4.0*np.pi*vd*time/lamda)
    model2 = pr*np.exp(-time*wd*2*np.pi/lamda)*np.sin(4.0*np.pi*vd*time/lamda)

    return np.sqrt( ( (re-model1)/re_error )**2 + ( (im-model2)/im_error )**2 )


#With a record dictionary from pydmap fit the contents
def lmfit2(record):
    import numpy as np
    from datetime import datetime
    from lmfit import Minimizer, Parameters

    now = datetime.now()

    #first get globally used values
    tfreq = record['tfreq'] * 1000.0
    mpinc = record['mpinc'] / 1000000.0
    smsep = record['smsep'] / 1000000.0
    lagfr = record['lagfr'] / 1000000.0
    nrang = record['nrang']
    ltab = record['ltab']
    mplgs = record['mplgs']
    nave = record['nave']
    acfd = record['acfd']
    lag0_power = np.array(record['pwr0'])

    #ignore the second lag 0 in the lag table
    ltab = ltab[0:-1]

    fit_record = build_fit_record(record)

    c = 299792458.0
    lamda = c/tfreq
    k = 2*np.pi/lamda
    nyquist_velocity = lamda/(4.*mpinc)

    lags=[]
    for pair in ltab:
        lags.append(pair[1]-pair[0])
    lags.sort()
    max_lag = lags[-1]

    t = np.array(lags)*mpinc
    ranges = np.arange(0,nrang)

    #Estimate the noise
    noise = estimate_noise(lag0_power)

    #Setup the fitted parameter lists
    fitted_power = list()
    fitted_width = list()
    fitted_vels = list()
    fitted_phis = list()
    fitted_power_e = list()
    fitted_width_e = list()
    fitted_vels_e = list()
    fitted_phis_e = list()
    slist = list()

    #next, iterate through all range gates and do fitting
    for gate in ranges:
        print("Beam: %s  Range Gate: %s" % (str(record['bmnum']),str(gate)))
        re = [x[0] for x in acfd[gate]]
        im = [x[1] for x in acfd[gate]]
        time = list(t)

        #estimate the upper limit of the self clutter
        clutter = list(estimate_selfclutter(nrang,ltab,smsep,mpinc,lagfr/2,gate,lag0_power)[lags])

        #find lags blanked due to Tx and identify "good" lags
        blanked = determine_tx_blanked(nrang,ltab,smsep,mpinc,lagfr/2,gate)
        blank_lags = [0 if len(blanked[x])==0 else 1 for x in blanked]
        
        #don't include any lags that are blanked
        j=0
        for i,bl in enumerate(blank_lags):
            if (bl == 1):
                re.pop(i-j)
                im.pop(i-j)
                time.pop(i-j)
                clutter.pop(i-j)
                j += 1
        re = np.array(re)
        im = np.array(im)
        time = np.array(time)
        clutter = np.array(clutter)

        #Now fit each acf using first order errors
        first_error = first_order_errors(lag0_power[gate],noise,clutter,nave)

        #set up the fitter and fit for the first time
        init_vels = np.linspace(-nyquist_velocity/2.,nyquist_velocity/2.,num=30)
        outs = list()

        for vel in init_vels:
            params = Parameters()
            params.add('power', value=lag0_power[gate],min=0)
            params.add('width', value=200.0,min=-100) #Need minimum, to stop magnitude model from diverging to infinity
            params.add('velocity', value=vel, min=-nyquist_velocity/2., max=nyquist_velocity/2.)

            minner = Minimizer(acf_residual, params, fcn_args=(time, re, im, first_error, first_error, lamda))
            outs.append(minner.minimize())

        chi2 = np.array([out.chisqr for out in outs])
        ind = np.where(chi2 == np.min(chi2))[0][0]

        pwr_fit = outs[ind].params['power'].value
        wid_fit = outs[ind].params['width'].value
        vel_fit = outs[ind].params['velocity'].value

        #Now get proper errorbars using fitted parameters and model
        acf_model = pwr_fit*np.exp(-time*2.*np.pi*wid_fit/lamda)*np.exp(1j*4.*np.pi*vel_fit*time/lamda)
        mag_model = np.abs(acf_model)
        rho = mag_model/mag_model[0]
        for i in range(len(rho)):
            if (rho[i] > 0.999):
                rho[i] = 0.999
            rho[i] = rho[i] * pwr_fit / (pwr_fit + noise + clutter[i])

        rho_re = rho*np.cos(4.*np.pi*vel_fit*time/lamda)
        rho_im = rho*np.sin(4.*np.pi*vel_fit*time/lamda)

        re_error = acf_error(pwr_fit,noise,clutter,nave,rho,rho_re)
        im_error = acf_error(pwr_fit,noise,clutter,nave,rho,rho_im)

        #Now second LMFIT
        outs2 = list()

        for vel in init_vels:
            params = Parameters()
            params.add('power', value=pwr_fit,min=0)
            params.add('width', value=wid_fit,min=-100)
            params.add('velocity', value=vel, min=-nyquist_velocity/2., max=nyquist_velocity/2.)

            minner = Minimizer(acf_residual, params, fcn_args=(time, re, im, re_error, im_error, lamda))
            outs2.append(minner.minimize())

        chi2 = np.array([out.chisqr for out in outs2])
        ind = np.where(chi2 == np.min(chi2))[0][0]

        pwr_fit = outs2[ind].params['power'].value
        wid_fit = outs2[ind].params['width'].value
        vel_fit = outs2[ind].params['velocity'].value

        # TO DO, implement errors that compare relative chi2 of the
        # multiple minima found. Just like in the C version.
        pwr_e = outs2[ind].params['power'].stderr
        wid_e = outs2[ind].params['width'].stderr
        vel_e = outs2[ind].params['velocity'].stderr

        #Now save fitted quantities into array
        slist.append(gate)
        fitted_power.append(pwr_fit)
        fitted_width.append(wid_fit)
        fitted_vels.append(vel_fit)
        fitted_power_e.append(pwr_e)
        fitted_width_e.append(wid_e)
        fitted_vels_e.append(vel_e)

    time_diff = (datetime.now()-now).total_seconds()
    print("It took %s seconds to fit one beam." % str(time_diff))

    #set ground scatter flags
    gflg = list()
    p_l = list()
    p_l_e = list()
    for i in range(len(slist)):
        if (np.abs(fitted_vels[i])-(30.-1./3.*np.abs(fitted_width[i])) < 0.):
            gflg.append(1)
        else:
            gflg.append(0)
        p_l.append(10.0*np.log10(fitted_power[i]/noise))
        p_l_e.append(10.0*np.log10((fitted_power_e[i]+fitted_power[i])/noise)-10.0*np.log10(fitted_power[i]/noise))

    #construct the fitted data dictionary that will be written to the fit file
    fit_record['slist'] = np.array(slist,dtype=np.int16)
    fit_record['nlag'] = mplgs * np.ones(len(slist),dtype=np.int16)
    fit_record['qflg'] = [1]*len(slist)
    fit_record['gflg'] = gflg
    fit_record['p_l'] = np.array(p_l,dtype=np.float32)
    fit_record['p_l_e'] = np.array(p_l_e,dtype=np.float32)
    fit_record['noise.sky'] = noise
    fit_record['noise.search'] = noise
    fit_record['noise.mean'] = noise
#    fit_record['p_s']
#    fit_record['p_s_e']
    fit_record['v'] = np.array(fitted_vels,dtype=np.float32)
    fit_record['v_e'] = np.array(fitted_vels_e,dtype=np.float32)
    fit_record['w_l'] = np.array(fitted_width,dtype=np.float32)
    fit_record['w_l_e'] = np.array(fitted_width_e,dtype=np.float32) 
#    fit_record['w_s'] = 
#    fit_record['w_s_e'] =
#    fit_record['sd_l'] = 
#    fit_record['sd_s'] = 
#    fit_record['sd_phi'] = 
#    fit_record['x_qflg'] = 
#    fit_record['x_gflg'] = 
#    fit_record['x_p_l'] = 
#    fit_record['x_p_l_e'] = 
#    fit_record['x_p_s'] = 
#    fit_record['x_p_s_e'] = 
#    fit_record['x_v'] = 
#    fit_record['x_v_e'] = 
#    fit_record['x_w_l'] = 
#    fit_record['x_w_l_e'] = 
#    fit_record['x_w_s'] = 
#    fit_record['x_w_s_e'] = 
#    fit_record['phi0'] = 
#    fit_record['phi0_e'] = 
#    fit_record['elv'] = 
#    fit_record['elv_low'] = 
#    fit_record['elv_high'] = 
#    fit_record['x_sd_l'] = 
#    fit_record['x_sd_s'] = 
#    fit_record['x_sd_phi'] = 

    return fit_record


#Main program loop. Opens rawacf, fits it, writes fitted
#data to output filename.
def main(input_filename, output_filename):

    import numpy as np

    fitted_records = list()

    #first, read the file we want to fit
    records = read_file(input_filename)

    #next iterate through all records and fit them
    for i,record in enumerate(records):
        print('Processing Record: %s' % str(i))
        fit_record = lmfit2(record)
        if isinstance(fit_record,dict):
            fitted_records.append(fit_record)

    #finally, write the fitted records to a dmap file
    write_file(output_filename, fitted_records)


def lmfit2_mp(pos,record):
    return pos,lmfit2(record)


# Code to run a test!
if __name__ == '__main__':
    import multiprocessing as mp

    pool = mp.Pool(2)

    input_filename = 'tests/test.rawacf'
    output_filename = 'tests/output.lmfit2'

    fitted_records = list()

    #first, read the file we want to fit
    records = read_file(input_filename)

    temp = [pool.apply_async(lmfit2_mp,args=(x,records[x])) for x in range(len(records))]

    fitted_records = [p.get() for p in temp]
    fitted_records.sort()
    fitted_records = [r[1] for r in fitted_records]

    write_file(output_filename, fitted_records)
